### PR TITLE
SparsityPattern refactoring

### PR DESCRIPTION
**Pull request by roystgnr**
This is a prerequisite for some more Isogeometric Analysis work, but that's conceptually separate (and not quite done) so it'll be in a later PR.  Hopefully the code quality improvements here (more modularization, less redundancy) stand on their own.  Also hopefully they work on their own, and didn't get shredded by the careful git add --patch fiddling that put them here.

### CODE REVIEW

File: include/base/sparsity_pattern.h [hunk 1]
```diff
@@ -80,8 +100,57 @@ static void sort_row (const BidirectionalIterator begin,
...
 class Build : public ParallelObject
 {
+public:
+  Build (const DofMap & dof_map_in,
+         const CouplingMatrix * dof_coupling_in,
+         const std::set<GhostingFunctor *> & coupling_functors_in,
+         const bool implicit_neighbor_dofs_in,
+         const bool need_full_sparsity_pattern_in);
+
+  Build (Build & other, Threads::split);
```
Comment (jwpeterson) [inline include/base/sparsity_pattern.h:110]: Consider defaulting the move constructor/deleting the other special functions for `Build` class? Not really necessary, but I feel it's good to have simply as documentation.

**Review by jwpeterson — APPROVED** [review]
Looks like a good refactoring to me... I couldn't tell from your last commit, what actually fixed the segfault we were seeing in that RB example?

**Comment by roystgnr** [comment]
I hadn't updated the implementation of DofMap::attach_matrix() to stay compatible with the SparseMatrix::_sp change.  The System-internal matrices that went through init_matrices() were still fine, so most examples worked, and external matrices where I was explicitly attaching a new SparsityPattern in my new code worked, but the previous method where you set up external matrices by attaching them to the DofMap, that one I broke.

File: include/base/sparsity_pattern.h [hunk 2]
```diff
@@ -107,6 +107,20 @@ class Build : public ParallelObject
          const bool implicit_neighbor_dofs_in,
          const bool need_full_sparsity_pattern_in);
 
+  /**
+   * Special functions.
+   * - The Build object holds references to DofMap and coupling
+   *   functors, therefore it can't be move-assigned.
+   */
+  Build (const Build &) = default;
+  Build & operator= (const Build &) = default;
```
Comment (jwpeterson) [inline include/base/sparsity_pattern.h:116]: You can't default copy-assign objects with reference members either, as that would be re-seating a reference. Copy-construct should be OK though. Some compilers correctly warn about this, but others don't.

**Pull request merged**